Undo Feature.

scofalik opened the pull request:
Undo Feature. Fixes #3.

pjasiun commented on src/undocommand.js:
```diff
@@ -0,0 +1,89 @@
+			}
+		}
+
+		this.fire( 'undo', undoBatch );
```
Because redo is also an `UndoCommand ` the event should be called `reverted`. It is strange that redo command fire `undo` event.

pjasiun commented on src/undocommand.js:
```diff
@@ -0,0 +1,89 @@
+	 * If not set, or invalid, the last added batch will be reverted and removed.
+	 */
+	_doExecute( batchIndex ) {
+		batchIndex = this._batchStack[ batchIndex ] ? batchIndex : this._batchStack.length - 1;
```
Comment needed.

scofalik replied:
What kind of comment? What is not clear in that line of code?

pjasiun replied:
I do not remember what was unclear, so nevermind.

pjasiun commented on src/undocommand.js:
```diff
@@ -0,0 +1,89 @@
+	 * @param {Number} [batchIndex] If set, batch under the given index on the stack will be reverted and removed.
+	 * If not set, or invalid, the last added batch will be reverted and removed.
+	 */
+	_doExecute( batchIndex ) {
```
How it is possible to get a batch index? If it not simple maybe batch instance should be a parameter?

scofalik replied:
The default undo implemented here will be (mostly?) used without setting this parameter. If a feature extends `UndoFeature` it could implement UI that will "let you choose" a `batchIndex` and fire `_doExecute` with given `batchIndex`.

In other words, let's say we have `SelectiveUndoManager` that extends `UndoManager` and introduces a special UI - a simple list of changes. Each list item is clickable and remembers its `batchIndex`. When it is clicked, it fires something like `editor.getCommand( 'selectiveUndo', this.batchIndex )`.

pjasiun replied:
You are talking here about a very theoretical API which may need this parameter. I think that it should be introduce when needed because we are not sure what is needed in fact.

pjasiun replied:
Also note that this is private method so no other class, including derived, should use it.

scofalik replied:
I know, but my example does not use it.

pjasiun commented on src/undofeature.js:
```diff
@@ -0,0 +1,82 @@
+		 * @private
+		 * @member {Set} undo.UndoFeature#_batchRegistry
+		 */
+		this._batchRegistry = new Set();
```
Why not `WeakSet`?

scofalik replied:
Changed

pjasiun replied:
I do not see the change.

pjasiun commented:
Beside these small changes mentioned above, two biggest missing parts are: selection handling and UI. If you do not want to handle them as a part of this ticket please create follow-ups. On the other hand selection may change undo so much that there is no point in reviewing undo without selection.

fredck commented:
> there is no point in considering undo without selection

Agree... selection is in fact one of the essential things related to undo.

scofalik commented:
I've implemented selection handling. It should work really well / perfect with non-selective undo where batches are not divided by other batches or when mixed delta is not conflicting with deltas from undone batch.

I did expanded implementation of range transformation so I expect that the selection should work okay in selective undo or mixed conflicting deltas too, but just don't expect wonders.

This comment is more suitable for ckeditor5-engine repo because that's where the magic happens, but I wanted to sum up the work here aswell.

Changes that are needed for this branch to work are on `ckeditor5-engine` branch `t/389`.

pjasiun commented on tests/undocommand.js:
```diff
@@ -0,0 +1,310 @@
+		} );
+	} );
+
+	describe( '_doExecute', () => {
```
There should not be tests for private class. If you want to test this method you should change it to "protected".

scofalik replied:
In fact, what should be done is:

* `_doExecute` should be `private` as it is called by `._execute`
* `._execute` should be tested.

I'll change it.

pjasiun commented on src/undocommand.js:
```diff
@@ -0,0 +1,192 @@
+		 * @private
+		 * @member {WeakMap.<engine.treeModel.Range>} undo.UndoCommand#_batchSelection
+		 */
+		this._batchSelection = new WeakMap();
```
If you keep ranges, instead of selections won't you loose direction after the undo?

scofalik replied:
This is true. I will add direction parameter too. I just don't want to keep whole `Selection` objects, which, I hope, is understandable.

pjasiun replied:
It is not, TBH. Group of ranges with a directions is exactly a selection, so why do not simple keep it?

scofalik replied:
I do not want to keep the selection because it stores more information (i.e. attributes), it mixes `EmitterMixin`, it may grow and get some default events. We do not want to store `Selection` but values needed to recreate it. Keeping selection makes all the process more complicated. Instead saving just ranges + one flag you, additionally, have to:
1. Create a new `Selection` instance.
2. When restoring selection, get the `Selection` instance, get it's ranges.
This makes script more complicated and for what reason? I'd understand if we would restore the whole selection but you can't even say that "`Selection` may change so it's better to keep the instance" cause all we operate on are ranges and that's what we care about. So even if `Selection` changes, we will still just use `.setRanges()` method.

scofalik replied:
Also keep in mind, that `Selection` uses `LiveRange`s so it would update whenever document is updated which is not what we want.

scofalik commented:
i added selection direction restoring. I've changed tests so they test `protected` `_execute` method instead of `private` `_doExecute`. I also removed a test for `addBatch` method as all it tested was `private` members.

I hope it is okay to merge now.

pjasiun commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+			batch,
+			{
+				ranges: Array.from( this.editor.document.selection.getRanges() ),
+				isBackward: this.editor.document.selection.isBackward
```
Still: why now `selection.clone()`?

scofalik replied:
I've already written why in one of comments (it got outdated and hidden). `Selection` uses `LiveRange`s and would get transformed if we saved `Selection`. This, and other reasons.

Reinmar commented on src/undofeature.js:
```diff
@@ -0,0 +1,82 @@
+ *
+ * Undo features brings in possibility to undo and re-do changes done in Tree Model by deltas through Batch API.
+ */
+export default class UndoFeature extends Feature {
```
The file and the class should be called just `Undo`.

scofalik replied:
Will we call bold feature `Bold`, etc.? I fear for some name collisions and having a `Feature` suffix makes it clear what it is.

Reinmar replied:
The idea is that if a features is called like the whole package it's enough to say `'undo'` to enable it (the name is extended automatically to `undo/undo`).

As for collisions, all other Bold related things must be named differently. Command will be `BoldCommand`, we may have `BoldSchema` or whatever else comes to our minds. The entry point of such a package is the feature so all other names have to be adjusted.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Undo command stores batches in itself and is able to and apply reverted versions of them on the document.
+ *
+ * undo.UndoCommand
```
Should be `@memberOf undo`.

scofalik replied:
Fixed. I don't know why I've written this :P

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+ */
+export default class UndoCommand extends Command {
+	/**
+	 * @see engine.command.Command
```
The whole comment isn't needed (hopefully) if you extend another class and overrides one of existing methods.

scofalik replied:
Removed comment

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+	 * Stores a batch in the command. Stored batches can be then reverted.
+	 *
+	 * @param {engine.treeModel.Batch} batch Batch to add.
+	 * @param {Array.<engine.treeModel.Range>} selectionRanges All ranges that were in the selection when batch got
```
To be removed.

scofalik replied:
Removed.

Reinmar commented on src/undofeature.js:
```diff
@@ -0,0 +1,82 @@
+		} );
+	}
+
+	destroy() {
```
This will be done by the `Plugin` class – https://github.com/ckeditor/ckeditor5/issues/114. So you can remove this method from here.

scofalik replied:
Removed

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+			this.editor.document.selection.setRanges( transformedRanges, selectionState.isBackward );
+		}
+
+		this.fire( 'undo', undoBatch );
```
It's a bit strange that the undo command fires the `undo` event. Especially when it's `redoCommand.on( 'undo', ... )`. So maybe we could find some better name. Like `revertedBatch`?

scofalik replied:
I agree but I don't find `revertedBatch` to be a good name. I will think of a better one.

scofalik replied:
Changed to `revert` event. Also I forgot about documenting that event.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+		 * @private
+		 * @member {Array.<engine.treeModel.Batch>} undo.UndoCommand#_batchStack
+		 */
+		this._batchStack = [];
```
Do we have stack limit implemented already?

scofalik replied:
Nope. I wasn't even sure how to get to editor's config correctly, so I left that thing for now.

Reinmar replied:
Reported in #4.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+		 * @private
+		 * @member {WeakMap} undo.UndoCommand#_batchSelection
+		 */
+		this._batchSelection = new WeakMap();
```
Name of this property sounds like a singular batch selection. BTW. Can't we keep them together with batches in the `_batchStack`? There's nothing wrong with using the WeakMap currently, but if we'll want to keep more information for each batch, we'll need more maps like this.

scofalik replied:
This implementation went through a few iterations and at one point I had to have it in `WeakMap` cause the selection for given batch might get updated. At this stage, I don't need to recognize selection state by `Batch` instance, so as you said, it can be merged with `_batchStack`.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+	/**
+	 * Checks whether this command should be enabled. Command is enabled when it has any batches in its stack.
+	 *
+	 * @private
```
It's documented as a protected method in the `Command`.

Reinmar replied:
Actually, the whole comment is unnecessary, because this method is documented in the parent class.

scofalik replied:
Removed

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+	 *
+	 * Fires `undo` event with reverted batch as a parameter.
+	 *
+	 * @private
```
Again, `@protected`.

scofalik replied:
I belive this is a mistake in `Command` class. `_doExecute` should be `private` and should be fired only by `_execute`. Take a look at `AttributeCommand`. If you agree that this is a mistake, I'll change it in `ckeditor5` repo.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,200 @@
+
+			// Then we check if two consecutive ranges are touching. We can do it pair by pair
+			// in one dimensional loop because ranges are sorted.
+			for ( let i = 1 ; i < transformed.length; i++ ) {
```
I just wanted to leave some comment in the middle of this function, so it looks like we've read and understood that code.

scofalik replied:
It's fairly simple, I can explain ...

Reinmar commented on src/undofeature.js:
```diff
@@ -0,0 +1,82 @@
+	 */
+	init() {
+		// Create commands.
+		this._redoCommand = new UndoCommand( this.editor );
```
I think that you can and perhaps even should do the initialization in the constructor. OTOH, I see that I'm doing initialization in delete feature in `init()` as well... I guess we need some rule of thumb here.

scofalik replied:
Aren't there any problems with having access to editor configuration or plugins loaded on demand? I have a feeling that we already discussed it and agreed that initialization other than settings values of class properties should be done in `init`.

scofalik commented:
Hey I've just fixed, and it is ready, I pushed a commit, so merge it maybe?

Reinmar commented:
BTW. which issue is this PR closing?

scofalik commented:
#3

Reinmar commented on src/undo.js:
```diff
@@ -0,0 +1,80 @@
+		} );
+
+		// Whenever batch is reverted by undo command, add it to redo history.
+		this._undoCommand.listenTo( this._redoCommand, 'revert', ( evt, batch ) => {
```
Shouldn't be the feature listening? In 99% of situation we should be using `listenTo()` on `this`, because the class which adds a listener is interested in removing it while it's gone.

scofalik replied:
You are right. I probably thought that this solution looks _cool_ but it will be better if feature is listening.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,195 @@
+		this._batchStack = [];
+	}
+
+	_checkEnabled() {
```
Misses docs.

Reinmar replied:
Oops, it's inherited from `Command`. Hmm... so maybe, to avoid confusion, we could use `@inheritDoc` in such cases. Then it will be clear that the doc comes from the parent.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,195 @@
+	 * @param {Number} [batchIndex] If set, batch under the given index on the stack will be reverted and removed.
+	 * If not set, or invalid, the last added batch will be reverted and removed.
+	 */
+	_doExecute( batchIndex ) {
```
We haven't thought about this but we need to document somewhere (public) the command's params. Or we can assume that the doc builder will somehow handle that.

Second thing is that I agree with @pjasiun that `batchIndex` is not a right param since you cannot retrieve it from anywhere. You don't even in what it is an index. Can't we pass there the batch itself and find the index?

pjasiun replied:
I still believe it should be `batch` since there is no simple way to get `batchIndex`.

scofalik replied:
Okay, you convinced me.

Reinmar commented on src/undocommand.js:
```diff
@@ -0,0 +1,195 @@
+	 * Executes the command: reverts a {@link engine.treeModel.Batch batch} added to the command's stack,
+	 * applies it on the document and removes the batch from the stack.
+	 *
+	 * @private
```
It's @protected.

pjasiun replied:
`_doExecute` is protected. Otherwise base class should not call it.

scofalik replied:
`_execute` calls `_doExecute` that's why I made `_doExecute` `private` but yes, you are right, if it is `private` it can't be called by parent class.

Reinmar commented:
I'm missing some more integrational tests, like:

1. Use composer to change the doc.
2. Check commands' states.
3. Undo and see that the data and selection are correct.

Those tests should be written with our test utils, so they are easy to read.

Having some low level tests is fine, but in a long run they will be hard to maintain, read and extend. That's when the easy-to-write and easy-to-read higher level tests are useful. I'd add some now (in a separate test file most likely) to have easier start in the future.

scofalik commented:
I could create some integrational tests using model test utils and see if model is fine but right now composer only has `deleteContents` so I'd have to create `batch`es by hand anyway, just like in `UndoCommand` tests. I would just not have to add them to `UndoCommand` instance. Same with selection, I'd have to manually set it.

I looked at `utils/src/diff.js` and `utils/src/difftochanges.js` and those maybe could be used to automatically generate batches.

Reinmar commented:
Doing changes is one part of a test. Second is setting data and checking the output. Currently you're building a document by creating lots of deltas, then do some operations (or even not – some tests starts with `undo._execute()`) and then some checks. This is fine (I guess), but it's not something that you can read and extend quickly.

By doing simple:

```js
setData( doc, '<p>foo[bar]</p>' );
doc.composer.deleteContents( doc.batch(), doc.selection ); // You can also use some deltas.
expect( getData( doc ), '<p>foo[]</p>' );
```

You will be able to quickly generate more test cases which underneath do quite similar things as you did with the low level API. 

E.g. let's imagine that we find a bug in a very specific case in the future. The first thing to do will be to code a test for it exactly like I showed. So we'll be definitely adding those. I'd just like to have some already, so I can verify on couple of cases that the undo performs as expected. I'm not able to do this quickly based on the existing tests.

scofalik commented:
I've added integration tests. Unfortunately, some of them fail, but not directly because of undo algorithms. I've commented them. I will create followup ticket for them in `ckeditor5-engine`.

scofalik commented:
I also fixed other issues you pointed out.

scofalik commented:
I commented out tests that failed and created followup issue in `ckeditor5-engine`

Reinmar commented:
I reported https://github.com/ckeditor/ckeditor5-undo/issues/6 for the ignored tests.

Reinmar commented on src/undocommand.js:
```diff
@@ -54,13 +59,26 @@ export default class UndoCommand extends Command {
+		if ( !batch ) {
+			batchIndex = this._batchStack.length - 1;
+		} else {
+			for ( let i = 0; i < this._batchStack.length; i++ ) {
```
You could've used: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find. I'll fix it.

Reinmar commented on tests/undo.js:
```diff
@@ -0,0 +1,340 @@
+
+import deleteContents from '/ckeditor5/engine/treemodel/composer/deletecontents.js';
+
+before( () => {
```
No point in optimising performance here. `beforeEach()` will do.

Reinmar commented:
http://tests.ckeditor.dev:1030/build/amd/tests/undo/undo#adding%20and%20removing%20content%20multiple%20remove%20and%20undo is failing for me. Could you check?

scofalik commented:
It needed a change in `ckeidtor5-engine` that I had only locally. Already fixed by #414.
